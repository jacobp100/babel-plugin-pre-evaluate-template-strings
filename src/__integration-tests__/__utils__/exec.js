/* @flow */

import childProcess from 'child_process';
import path from 'path';

export function transpile(
  source: string,
  pluginOptions: any = { cache: false, extract: false },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_TRANSPILE_BIN = path.join(__dirname, './transpile.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_TRANSPILE_BIN} '${JSON.stringify(
      source
    )}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [transpiled, rawStyles] = stdout.toString().split('\n');
  const styles = JSON.parse(rawStyles || '{}');
  return {
    code: JSON.parse(transpiled || '""'),
    styles,
    getCSSForClassName(className: string) {
      return styles[`.${className}`];
    },
  };
}

export function extract(
  source: string,
  pluginOptions: any = { cache: false, extract: true },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_EXTRACT_BIN = path.join(__dirname, './extract.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_EXTRACT_BIN} '${JSON.stringify(source)}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [data, filenames, transpiled] = stdout.toString().split('\n');
  return {
    data: JSON.parse(data || '""'),
    filenames: JSON.parse(filenames || '[]'),
    transpiled: JSON.parse(transpiled || '""'),
  };
}
