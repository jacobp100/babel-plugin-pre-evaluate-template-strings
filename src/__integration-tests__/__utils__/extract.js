const fs = require('fs');

const writeFileSync = fs.writeFileSync;

let buffer = '';
const filenames = [];

fs.writeFileSync = (filename, data, opts) => {
  if (/\.css$/.test(filename)) {
    filenames.push(filename);
    buffer += data;
    return;
  }
  writeFileSync(filename, data, opts);
};

const transpile = require('./transpile');

const code = transpile(
  JSON.parse(process.argv[2]),
  JSON.parse(process.argv[3])
);

console.log(JSON.stringify(buffer));
console.log(JSON.stringify(filenames));
console.log(JSON.stringify(code || ''));
