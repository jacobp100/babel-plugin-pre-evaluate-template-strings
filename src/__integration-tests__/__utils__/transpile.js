const babel = require('babel-core');
const path = require('path');
const dedent = require('dedent');

const CSS_PATH = require.resolve(path.join(process.cwd(), 'build/css'));
const SHEET_PATH = require.resolve(path.join(process.cwd(), 'build/sheet'));
const PREVAL_PLUGIN_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/preval-extract')
);
const MODULE_SYSTEM_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/lib/moduleSystem')
);

function transpile(source, options = { pluginOptions: {}, babelOptions: {} }) {
  return babel.transform(
    dedent`
      import css from '${CSS_PATH}';

      ${source}
    `,
    Object.assign(
      {
        presets: ['env', 'stage-2', 'react'],
        plugins: [[PREVAL_PLUGIN_PATH, options.pluginOptions]],
        sourceMaps: false,
      },
      options.babelOptions
    )
  ).code;
}

module.exports = transpile;

if (require.main && module.filename === require.main.filename) {
  const code = transpile(
    JSON.parse(process.argv[2]),
    JSON.parse(process.argv[3])
  );

  if (code) {
    console.log(JSON.stringify(code));

    // eslint-disable-next-line global-require
    const getCachedModule = require(MODULE_SYSTEM_PATH).getCachedModule;
    const sheet = getCachedModule(SHEET_PATH);
    const styles = sheet
      ? getCachedModule(SHEET_PATH).exports.default.styles()
      : {};
    console.log(JSON.stringify(styles || ''));
  }
}
