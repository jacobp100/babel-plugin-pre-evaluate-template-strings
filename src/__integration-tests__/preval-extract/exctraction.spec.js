/* eslint-disable no-template-curly-in-string */
/* @flow */

import path from 'path';
import dedent from 'dedent';
import { extract } from '../__utils__/exec';

describe('preval-extract babel plugin with extraction enabled', () => {
  it('should extract all styles to a single file', () => {
    const filename = path.join(process.cwd(), 'test.js');
    const { data: data1, filenames: filenames1 } = extract(
      dedent`
        const header = css\`
          font-size: 3em;
        \`;
        `,
      { single: true },
      { filename }
    );

    const { data: data2, filenames: filenames2 } = extract(
      dedent`
        const body = css\`
          font-weight: bold;
        \`;
        `,
      { single: true },
      { filename }
    );

    expect(filenames1).toEqual(filenames2);
    expect(data1).toMatchSnapshot();
    expect(data2).toMatchSnapshot();
  });

  it('should extract each style to separate file and include it into source file', () => {
    const filename1 = path.join(process.cwd(), 'test1.js');
    const filename2 = path.join(process.cwd(), 'test2.js');

    const {
      data: data1,
      filenames: filenames1,
      transpiled: transpiled1,
    } = extract(
      dedent`
        const header = css\`
          font-size: 3em;
        \`;
        `,
      undefined,
      { filename: filename1 }
    );

    const {
      data: data2,
      filenames: filenames2,
      transpiled: transpiled2,
    } = extract(
      dedent`
        const body = css\`
          font-weight: bold;
        \`;
        `,
      undefined,
      { filename: filename2 }
    );

    expect(transpiled1).toMatch(`require('./.linaria-cache/test1.css')`);
    expect(transpiled2).toMatch(`require('./.linaria-cache/test2.css')`);

    expect(data1).toMatchSnapshot();
    expect(data2).toMatchSnapshot();
    expect(filenames1).toEqual([
      path.join(process.cwd(), '.linaria-cache/test1.css'),
    ]);
    expect(filenames2).toEqual([
      path.join(process.cwd(), '.linaria-cache/test2.css'),
    ]);
  });

  it('extract styles to a given file', () => {
    const filename = path.join(process.cwd(), 'test.js');
    const { data, filenames } = extract(
      dedent`
        const header = css\`
          font-size: 3em;
        \`;
        `,
      { single: true, filename: 'styles-static.css', cache: false },
      { filename }
    );

    expect(filenames).toEqual([
      path.join(path.dirname(filename), '.linaria-cache/styles-static.css'),
    ]);
    expect(data).toMatchSnapshot();
  });

  it('extract styles to a given file with output directory specified', () => {
    const filename = path.join(process.cwd(), 'test.js');
    const { data, filenames } = extract(
      dedent`
        const header = css\`
          font-size: 3em;
        \`;
        `,
      {
        single: true,
        filename: 'styles-static.css',
        outDir: 'output',
        cache: false,
      },
      { filename }
    );

    expect(filenames).toEqual([
      path.join(path.dirname(filename), 'output/styles-static.css'),
    ]);
    expect(data).toMatchSnapshot();
  });
});
