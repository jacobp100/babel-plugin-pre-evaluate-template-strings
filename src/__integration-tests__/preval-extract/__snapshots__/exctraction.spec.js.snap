// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract babel plugin with extraction enabled extract styles to a given file 1`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */
._header__zkmz8m{font-size:3em;}"
`;

exports[`preval-extract babel plugin with extraction enabled extract styles to a given file with output directory specified 1`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */
._header__zkmz8m{font-size:3em;}"
`;

exports[`preval-extract babel plugin with extraction enabled should extract all styles to a single file 1`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */

._header__zkmz8m{font-size:3em;}"
`;

exports[`preval-extract babel plugin with extraction enabled should extract all styles to a single file 2`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */

._body__zkmz8m{font-weight:bold;}"
`;

exports[`preval-extract babel plugin with extraction enabled should extract each style to separate file and include it into source file 1`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */
._header__jz3lbe{font-size:3em;}"
`;

exports[`preval-extract babel plugin with extraction enabled should extract each style to separate file and include it into source file 2`] = `
"/* THIS FILE IS AUTOGENERATED. DO NOT EDIT IT DIRECTLY, NOR COMMIT IT TO VERSION CONTROL. */
._body__16a4ra6{font-weight:bold;}"
`;
